ndcube/ndcube.py

```diff
@@ -399,6 +399,42 @@ def __init__(self, data, wcs=None, uncertainty=None, mask=None, meta=None,
                 global_coords = deepcopy(global_coords)
             self._global_coords = global_coords
 
+    @property
+    def data(self):
+        """
+        `~numpy.ndarray`-like : The stored dataset.
+
+        Notes
+        -----
+        It is possible to set the ``.data`` attribute on a `NDCube` with an
+        array-like object of the same shape. However, this is really only
+        intended for replacing the data with a different object representing
+        the same physical data as no other properties of the cube will be
+        changed, such as uncertainty or unit.
+        """
+        return super().data
+
+    @data.setter
+    def data(self, value):
+        # In an array-agnostic way check the shape is the same
+        if not hasattr(value, "shape") or value.shape != self.data.shape:
+            raise TypeError(f"Can only set data with an array-like object of the same shape ({self.data.shape})")
+
+        # Other masked arrays are hard to detect reliably
```

Comment:
So what happens if `value` a different type of masked array?  Does `.data` become that type of masked array?

Reply:
yes

Reply:
Probably the same thing.

Reply:
yes, the same thing:
```
In [10]: NDCube(data=dask.array.ma.zeros_like(np.array((10,10))), wcs=WCS()).data
Out[10]: dask.array<<numpy.ma.core._convert2ma object at 0x7f8144495c5, shape=(2,), dtype=int64, chunksize=(2,), chunktype=numpy.MaskedArray>
```

this is just unsupported anywhere.